File: packages/sdk/src/resources/credit-cards.ts
import { ValidationError } from "../utils/errors";
import { BaseResource } from "./base";
import type {
  CreditCardComparisonRequest,
  CreditCardComparisonResponse,
} from "../types";

export class CreditCards extends BaseResource {
  async compare(
    request: CreditCardComparisonRequest,
  ): Promise<CreditCardComparisonResponse> {
    this.validateCardRequest(request);
    const toolResult = await this.client.callTool("find_credit_cards", request);
    const data =
      this.parseJsonResponse<Omit<CreditCardComparisonResponse, "widget">>(
        toolResult,
      );
    return {
      ...data,
      widget: this.getWidget(toolResult),
    };
  }

  private validateCardRequest(request: CreditCardComparisonRequest): void {
    if (
      !request.creditScore ||
      request.creditScore < 300 ||
      request.creditScore > 850
    ) {
      throw new ValidationError("Valid credit score (300-850) is required");
    }

    if (!request.monthlySpend || request.monthlySpend < 0) {
      throw new ValidationError("Monthly spend must be a positive number");
    }
  }
}

File: packages/python/securelend/resources/credit_cards.py
from ..utils.errors import ValidationError
from .base import BaseResource
from ..types import CreditCardComparisonRequest, CreditCardComparisonResponse

class CreditCards(BaseResource):
    async def compare(
        self, request: CreditCardComparisonRequest
    ) -> CreditCardComparisonResponse:
        self._validate_card_request(request)
        tool_result = await self._client.call_tool("find_credit_cards", request)
        data = self._parse_json_response(tool_result)
        data["widget"] = self._get_widget(tool_result)
        return data

    def _validate_card_request(self, request: CreditCardComparisonRequest) -> None:
        credit_score = request.get("creditScore")
        if not credit_score or not (300 <= credit_score <= 850):
            raise ValidationError("Valid credit score (300-850) is required")

        monthly_spend = request.get("monthlySpend")
        if monthly_spend is None or monthly_spend < 0:
            raise ValidationError("Monthly spend must be a positive number")
